src/app.js

```diff
@@ -1,48 +1,53 @@
-const express = require('express');
-const path = require('path');
-const ejsMate = require('ejs-mate');
-const session = require('express-session');
-const flash = require('connect-flash');
-const passport = require('passport');
-const LocalStrategy = require('passport-local');
-const methodOverride = require('method-override');
-const MongoStore = require('connect-mongo');
-const { secret } = require('./configs');
-const { healthcheckRoutes, v1Routes } = require('./routes');
-const User = require('./models/User');
-const { dbUrl } = require('./configs/database');
+const express = require("express");
+const path = require("path");
+const ejsMate = require("ejs-mate");
+const session = require("express-session");
+const flash = require("connect-flash");
+const passport = require("passport");
+const LocalStrategy = require("passport-local");
+const methodOverride = require("method-override");
+const MongoStore = require("connect-mongo");
+const { secret } = require("./configs");
+const { healthcheckRoutes, v1Routes } = require("./routes");
+const User = require("./models/User");
+const { dbUrl } = require("./configs/database");
+const imageUpload = require("./middleware/upload");
 
 const app = express();
 
-app.engine('ejs', ejsMate);
-app.set('view engine', 'ejs');
-app.set('views', path.join(__dirname, 'views'));
-app.use(express.static(path.join(__dirname, 'public')));
-app.use(methodOverride('_method'));
+app.engine("ejs", ejsMate);
+app.set("view engine", "ejs");
+app.set("views", path.join(__dirname, "views"));
+app.use(express.static(path.join(__dirname, "public")));
+app.use(methodOverride("_method"));
 
 const store = MongoStore.create({
-    mongoUrl: dbUrl,
-    ttl: 14 * 24 * 60 * 60,
-    autoRemove: 'native',
-    touchAfter: 24 * 3600, // time period in seconds
+  mongoUrl: dbUrl,
+  ttl: 14 * 24 * 60 * 60,
+  autoRemove: "native",
+  touchAfter: 24 * 3600, // time period in seconds
 });
 
 const sessionConfig = {
-    store,
-    secret,
-    resave: false,
-    saveUninitialized: true,
-    cookie: {
-        // secure: true
-        httpOnly: true,
-        maxAge: 1000 * 60 * 60 * 24 * 7,
-    },
+  store,
+  secret,
+  resave: false,
+  saveUninitialized: true,
+  cookie: {
+    // secure: true
+    httpOnly: true,
+    maxAge: 1000 * 60 * 60 * 24 * 7,
+  },
 };
 
 app.use(session(sessionConfig));
 app.use(flash());
-app.use(express.json({ limit: '10mb' }));
-app.use(express.urlencoded({ limit: '10mb', extended: true, parameterLimit: 5000 }));
+app.use(express.json({ limit: "10mb" }));
+app.use(
+  express.urlencoded({ limit: "10mb", extended: true, parameterLimit: 5000 })
+);
+// middleware to process multipart/form-data requests
+app.use(imageUpload);
```

Comment:
1. How we are making sure that this middleware is called only for create product request? 
2. And how cloudinary image url is passed and saved into the db

Reply:
1. This middleware is a multer middleware with cloudinary as the underlying storage layer. So the intent of this middleware is to upload any images by intercepting multipart/form-data to cloudinary. The ability to upload image is applicable to create product, edit product and can also be added to create/edit a review for a product.
2. The middleware is created using multer package only and thus after uploading an image to cloudinary, the middleware injects a file object in the request which contains the details of the file and the cloudinary URL generated for the image.

Reply:
1. This is actually bit confusing, can we add middleware to the route itself, adding such middleware at the top level might cause issue / confusion. lets add it at route level,  it will give more readability and will be specific to the intended route

Reply:
ok
I have added the uploadImage middleware to POST /products route for this change